feat(logger): add a simple logging module

ullasholla opened the pull request:
This PR adds a simple logging module. It uses [Winston](https://github.com/winstonjs/winston) to do the heavy-lifitng but the API is agnostic to Winston. This is done so that Winston could be replaced by another logging module in the future if desired. The clients of the library will continue to function without any changes. The [README](https://github.com/SAP/open-ux-tools/blob/feat/add-logger/packages/logger/README.md) document has more information about the module.

ullasholla commented:
Test coverage:
![image](https://user-images.githubusercontent.com/212316/148896251-9c48d7e2-49f3-41e4-b98c-c0663a6edf5d.png)

tobiasqueck reviewed (approved):
* well structured and easy to understand coding
* simple to use but also very flexible if specific configurations are needed
* I especially like the dev-null logging :)
* tried it for my use cases in the `axios-extension` and it worked without problems
* minor suggestion (but not required): user type `Partial<LoggerOptions> for the logger options property of the constructor so that I can also just provide a log level and not a transport (or vice versa)
* please add more js-doc descriptions to your functions

tobiasqueck commented:
@zdravko-georgiev can you check if this works with logging in ui5 middlewares and tasks?

zdravko-georgiev commented:
> @zdravko-georgiev can you check if this works with logging in ui5 middlewares and tasks?

@tobiasqueck Checking

zdravko-georgiev commented:
Is this how the `info` supposed to look like in the console?
![image (8)](https://user-images.githubusercontent.com/91968887/148915510-f2c09710-3b42-495f-afe0-64c802ff9b0b.png)

ullasholla commented:
> Is this how the `info` supposed to look like in the console? ![image (8)](https://user-images.githubusercontent.com/91968887/148915510-f2c09710-3b42-495f-afe0-64c802ff9b0b.png)

Yes. I can't fault you for not getting all way the way through the [README](https://github.com/SAP/open-ux-tools/blob/feat/add-logger/packages/logger/README.md#not-supported) :P

![image](https://user-images.githubusercontent.com/212316/148979737-c1943465-06c3-4824-a9fd-99ceafe68ce3.png)

ullasholla commented:
@IainSAP, @zdravko-georgiev, @tobiasqueck: I changed the default format for console to be plain text instead of JSON (it is colored and timestamped though. Piping the output will not use the [color encoding](https://en.wikipedia.org/wiki/ANSI_escape_code) - this is to facilitate both manual inspection and text manipulation (color encoding can be distracting and painful to work with)). `FileTransport` still uses JSON.

@IainSAP, I am slightly deviating from the format used in Fiori tools (levels are not uppercased and the log entry is not enclosed by []). Let me know if you have any concerns with these. (Not using uppercase as I use `format.colorize()` and it expects lowercased levels. The effort to keep both didn't seem worth it at this point).

An example logger that logs to console and two files simultaneously (the transports are added at different points in time and don't contain identical logs/levels):
![image](https://user-images.githubusercontent.com/212316/149013659-06005d72-cc29-455b-b7ca-071d5a1aa1d4.png)


Piping the output to a file (`sample.log` does not have any ANSI escape codes):
![image](https://user-images.githubusercontent.com/212316/149013774-959479b4-164d-45bb-ab89-bb02d55367d7.png)

@zdravko-georgiev, looks like the ui5-logger uses [`npmlog`](https://github.com/npm/npmlog) for logging. I'm still checking to see if the output can be made compatible.

zdravko-georgiev commented on packages/logger/src/winston-logger/vscode-output-channel-transport.ts:
```diff
@@ -0,0 +1,37 @@
+import Transport from 'winston-transport';
+// import { window, OutputChannel } from 'vscode';
```
Can we remove this?

zdravko-georgiev commented:
There is one empty file `packages/logger/foo`. Can we remove it?

IainSAP commented:
Hi Ullas,

How will this interact with the existing SAP vscode logging we use in App Gen. Currently we already have rolling file and channel logging provided by the SAP VSCode logger. We do not want multiple log files being created when generating apps for example, or have multiple output channels for different parts of generation. This would complicate the logging. Does the open source logging proposal support integration with the existing SAP VSCode logging being used?

-Iain


From: Ullas Holla ***@***.***>
Reply to: SAP/open-ux-tools ***@***.***>
Date: Tuesday 11 January 2022 at 20:17
To: SAP/open-ux-tools ***@***.***>
Cc: Iain O'Farrell ***@***.***>, Mention ***@***.***>
Subject: Re: [SAP/open-ux-tools] feat(logger): add a simple logging module (PR #257)


@IainSAP<https://eur03.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FIainSAP&data=04%7C01%7Ciain.o-farrell%40sap.com%7C0e8c2e0e431b40b7d56208d9d53f67cd%7C42f7676cf455423c82f6dc2d99791af7%7C0%7C0%7C637775290566087863%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000&sdata=PbRMLerJlgLRVaoE4P6%2FUG4CeewdHsnU49YolkiAvBg%3D&reserved=0>, @zdravko-georgiev<https://eur03.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fzdravko-georgiev&data=04%7C01%7Ciain.o-farrell%40sap.com%7C0e8c2e0e431b40b7d56208d9d53f67cd%7C42f7676cf455423c82f6dc2d99791af7%7C0%7C0%7C637775290566087863%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000&sdata=za%2Fiy4sjFZPlTf9zSUZZBUW1TTF2MhnmZYF9w4XTvIU%3D&reserved=0>, @tobiasqueck<https://eur03.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Ftobiasqueck&data=04%7C01%7Ciain.o-farrell%40sap.com%7C0e8c2e0e431b40b7d56208d9d53f67cd%7C42f7676cf455423c82f6dc2d99791af7%7C0%7C0%7C637775290566087863%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000&sdata=4DWkQFLHsWxH76ztgZu2uA5L11Ii1OBF4ycLKphCLMY%3D&reserved=0>: I changed the default format for console to be plain text instead of JSON (it is colored and timestamped though. Piping the output will not use the color encoding<https://eur03.safelinks.protection.outlook.com/?url=https%3A%2F%2Fen.wikipedia.org%2Fwiki%2FANSI_escape_code&data=04%7C01%7Ciain.o-farrell%40sap.com%7C0e8c2e0e431b40b7d56208d9d53f67cd%7C42f7676cf455423c82f6dc2d99791af7%7C0%7C0%7C637775290566087863%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000&sdata=Hc1%2FOSOgkCzs42b0qr1BomXz2YHxcpsndvU%2BbJX3e5c%3D&reserved=0> - this is to facilitate both manual inspection and text manipulation (color encoding can be distracting and painful to work with)). FileTransport still uses JSON.

@IainSAP<https://eur03.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FIainSAP&data=04%7C01%7Ciain.o-farrell%40sap.com%7C0e8c2e0e431b40b7d56208d9d53f67cd%7C42f7676cf455423c82f6dc2d99791af7%7C0%7C0%7C637775290566087863%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000&sdata=PbRMLerJlgLRVaoE4P6%2FUG4CeewdHsnU49YolkiAvBg%3D&reserved=0>, I am slightly deviating from the format used in Fiori tools (levels are not uppercased and the log entry is not enclosed by []). Let me know if you have any concerns with these. (Not using uppercase as I use format.colorize() and it expects lowercased levels. The effort to keep both didn't seem worth it at this point).

An example logger that logs to console and two files simultaneously (the transports are added at different points in time and don't contain identical logs/levels):
[image]<https://eur03.safelinks.protection.outlook.com/?url=https%3A%2F%2Fuser-images.githubusercontent.com%2F212316%2F149013659-06005d72-cc29-455b-b7ca-071d5a1aa1d4.png&data=04%7C01%7Ciain.o-farrell%40sap.com%7C0e8c2e0e431b40b7d56208d9d53f67cd%7C42f7676cf455423c82f6dc2d99791af7%7C0%7C0%7C637775290566087863%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000&sdata=XqjbNlNGGzNNA2UW%2BbokgXw7P%2B33UQl9Q3OjBqCIxt0%3D&reserved=0>

Piping the output to a file (sample.log does not have any ANSI escape codes):
[image]<https://eur03.safelinks.protection.outlook.com/?url=https%3A%2F%2Fuser-images.githubusercontent.com%2F212316%2F149013774-959479b4-164d-45bb-ab89-bb02d55367d7.png&data=04%7C01%7Ciain.o-farrell%40sap.com%7C0e8c2e0e431b40b7d56208d9d53f67cd%7C42f7676cf455423c82f6dc2d99791af7%7C0%7C0%7C637775290566087863%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000&sdata=M99YaXmwUv1bi4z2GhD7lpaLhZOc%2FerCNv0UXVpTE0Q%3D&reserved=0>

@zdravko-georgiev<https://eur03.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fzdravko-georgiev&data=04%7C01%7Ciain.o-farrell%40sap.com%7C0e8c2e0e431b40b7d56208d9d53f67cd%7C42f7676cf455423c82f6dc2d99791af7%7C0%7C0%7C637775290566087863%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000&sdata=za%2Fiy4sjFZPlTf9zSUZZBUW1TTF2MhnmZYF9w4XTvIU%3D&reserved=0>, looks like the ui5-logger uses npmlog<https://eur03.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnpm%2Fnpmlog&data=04%7C01%7Ciain.o-farrell%40sap.com%7C0e8c2e0e431b40b7d56208d9d53f67cd%7C42f7676cf455423c82f6dc2d99791af7%7C0%7C0%7C637775290566087863%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000&sdata=fTXUmm02bz56fHh%2FJXAtxEu%2FTLwORxRPA6NKOcyVhEw%3D&reserved=0> for logging. I'm still checking to see if the output can be made compatible.

—
Reply to this email directly, view it on GitHub<https://eur03.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FSAP%2Fopen-ux-tools%2Fpull%2F257%23issuecomment-1010326678&data=04%7C01%7Ciain.o-farrell%40sap.com%7C0e8c2e0e431b40b7d56208d9d53f67cd%7C42f7676cf455423c82f6dc2d99791af7%7C0%7C0%7C637775290566087863%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000&sdata=cSDpbMtkvP0H7fUP1jvlxuv92sHRiRLARvfwHUAStSA%3D&reserved=0>, or unsubscribe<https://eur03.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FALDBLRWJAVGNG4GW5HOECKDUVSF53ANCNFSM5LVQM5WQ&data=04%7C01%7Ciain.o-farrell%40sap.com%7C0e8c2e0e431b40b7d56208d9d53f67cd%7C42f7676cf455423c82f6dc2d99791af7%7C0%7C0%7C637775290566087863%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000&sdata=hEuiSFlQ7FtBUWvLuMmj2F2zqHm6frjgvyi0mM2zMnI%3D&reserved=0>.
Triage notifications on the go with GitHub Mobile for iOS<https://eur03.safelinks.protection.outlook.com/?url=https%3A%2F%2Fapps.apple.com%2Fapp%2Fapple-store%2Fid1477376905%3Fct%3Dnotification-email%26mt%3D8%26pt%3D524675&data=04%7C01%7Ciain.o-farrell%40sap.com%7C0e8c2e0e431b40b7d56208d9d53f67cd%7C42f7676cf455423c82f6dc2d99791af7%7C0%7C0%7C637775290566087863%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000&sdata=Z9194aBDBMEVyghyWupY0svEK51MnaCnr76v3dFzaX4%3D&reserved=0> or Android<https://eur03.safelinks.protection.outlook.com/?url=https%3A%2F%2Fplay.google.com%2Fstore%2Fapps%2Fdetails%3Fid%3Dcom.github.android%26referrer%3Dutm_campaign%253Dnotification-email%2526utm_medium%253Demail%2526utm_source%253Dgithub&data=04%7C01%7Ciain.o-farrell%40sap.com%7C0e8c2e0e431b40b7d56208d9d53f67cd%7C42f7676cf455423c82f6dc2d99791af7%7C0%7C0%7C637775290566087863%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000&sdata=Q3p6THjJKamHwFeDEGxRvxgyJd%2F1YAOqG36h%2Fdl7JCo%3D&reserved=0>.
You are receiving this because you were mentioned.Message ID: ***@***.***>

ullasholla commented:
> There is one empty file `packages/logger/foo`. Can we remove it?

That's there for good luck 😬

zdravko-georgiev commented on packages/logger/src/winston-logger/logger.ts:
```diff
@@ -39,6 +41,18 @@ export class WinstonLogger implements Logger {
+        if (typeof data === 'string') {
+            this._logger.log(this.winstonLevel, data);
+        } else {
+            this._logger.log(toWinstonLogLevel(data.level)!, data.message);
```
Is the `!` on purpose?

ullasholla replied:
Yes, it was on purpose but it's gone now.

ullasholla commented:
Child loggers are now supported. They add an prefix to the log to make it clear which component is doing the logging. The `FileTransport` will have the property `label` to filter the logs on, if required.

On the console, using child loggers looks like this:
![image](https://user-images.githubusercontent.com/212316/149395176-83ab6607-eed5-4d4b-ace0-d958f36dfc6b.png)

zdravko-georgiev reviewed (approved):
Code looks good, 2 Unit test are failing

tobiasqueck reviewed (approved):
Reviewed:

- [x] understandability of the code - nice, clean structure, however a little more complex than I expected 
- [x] very good test coverage
- [x] works with planned `axios-extension` module

tobiasqueck commented:
> Code looks good, 2 Unit test are failing

@zdravko-georgiev: not for me, which ones cause problems for you?

zdravko-georgiev commented:
> Details

@tobiasqueck The PR build is failing https://github.com/SAP/open-ux-tools/runs/4813000583?check_suite_focus=true

devinea commented on package.json:
```diff
@@ -6,6 +6,7 @@
     "devDependencies": {
         "@changesets/cli": "2.19.0",
         "@types/jest": "27.4.0",
+        "@types/node": "17.0.8",
```
Should opt for the lowest supported version of node (e.g. 12) and only LTS versions

ullasholla replied:
Fixed in https://github.com/SAP/open-ux-tools/pull/257/commits/c2348bd8d23c4b47495add3e57d3881c8eb5241f

devinea commented on packages/logger/.eslintrc:
```diff
@@ -0,0 +1,16 @@
+        "jsdoc/check-tag-names": 0,
+        "@typescript-eslint/no-unused-vars": ["error", { "varsIgnorePattern": "^_", "argsIgnorePattern": "^_" }]
+    }
+}
```
why have custom rules here. Shouldn't all modules use the same?

ullasholla replied:
Fixed in https://github.com/SAP/open-ux-tools/pull/257/commits/0ad205679050af6a49ee1e7619f136df131219a8

devinea commented on packages/logger/README.md:
```diff
@@ -0,0 +1,75 @@
+### Not supported
+The following options are not supported yet. There are no definite plans to add them in the future. They will be considered on a need-basis.
+
+* Child loggers
```
This has been added since?

ullasholla replied:
Good catch, thanks.

ullasholla replied:
Fixed.

devinea commented on packages/logger/package.json:
```diff
@@ -0,0 +1,41 @@
+        "lint": "eslint . --ext .ts",
+        "lint:fix": "eslint . --ext .ts --fix",
+        "test": "jest --maxWorkers=1 --colors",
+        "test-u": "jest --maxWorkers=1 --colors -u",
```
I recall you asked to remove `test-u` from the other modules

ullasholla replied:
I kept this to be consistent with the other modules. Shall I remove this from all the modules?

devinea replied:
may as well

ullasholla replied:
Changed my mind, don't want to make changes in the other modules here. Removed in logger, ta.

devinea commented on packages/logger/package.json:
```diff
@@ -0,0 +1,41 @@
+    },
+    "dependencies": {
+        "chalk": "4.1.2",
+        "lodash.clonedeep": "4.5.0",
```
```suggestion
        "lodash": "4.17.21"
```
please use `lodash` instead and `import cloneDeep from 'lodash/clonedeep';`
Makes for better treeshaking with webpack 🤷

ullasholla replied:
Fixed in https://github.com/SAP/open-ux-tools/pull/257/commits/34fab90d6a1a69d58b44c43d0be32b66836d41df

devinea commented on packages/logger/package.json:
```diff
@@ -0,0 +1,41 @@
+    },
+    "devDependencies": {
+        "@types/debug": "4.1.7",
+        "@types/lodash.clonedeep": "4.5.6",
```
```suggestion
        "@types/lodash": "4.14.176"
```

ullasholla replied:
Fixed in https://github.com/SAP/open-ux-tools/pull/257/commits/34fab90d6a1a69d58b44c43d0be32b66836d41df

devinea commented on packages/logger/src/types.ts:
```diff
@@ -0,0 +1,99 @@
+import cloneDeep from 'lodash.clonedeep';
```
```suggestion
import cloneDeep from 'lodash/clonedeep';
```

ullasholla replied:
Fixed in https://github.com/SAP/open-ux-tools/pull/257/commits/34fab90d6a1a69d58b44c43d0be32b66836d41df

devinea commented on packages/logger/tsconfig.json:
```diff
@@ -0,0 +1,12 @@
+        "sourceMap": true,
+        "baseUrl": "./src",
+        "resolveJsonModule": true,
+        "lib": ["ESNext"]
```
```suggestion
        "lib": ["2019"]
```
safer for node 12 support.

ullasholla replied:
Fixed in https://github.com/SAP/open-ux-tools/pull/257/commits/c2348bd8d23c4b47495add3e57d3881c8eb5241f

devinea commented on packages/logger/src/winston-logger/logger.ts:
```diff
@@ -0,0 +1,169 @@
+    // Maintain of map of transports. This is useful for adding/removing transports
+    protected transportMap: Map<Transport, WinstonTransport>;
+    // Cache child loggers by logPrefix. Use weak references to avoid causing memory leaks
+    protected childMap: Map<string, WeakRef<Logger>> = new Map();
```
It seems WeakRef is only supported from Node 14.6.0+. We need node 12 support for a few more months atleast.

ullasholla replied:
Removed the need for `WeakRef`. Fixed in https://github.com/SAP/open-ux-tools/pull/257/commits/c2348bd8d23c4b47495add3e57d3881c8eb5241f

ullasholla commented:
> @tobiasqueck The PR build is failing https://github.com/SAP/open-ux-tools/runs/4813000583?check_suite_focus=true

@zdravko-georgiev & @tobiasqueck, this is to do with Node 12 vs Node 14. Like @devinea pointed out, it's to do with `WeakRef` not being supported yet on Node 12 yet.

`WeakRef`s are used to keep to make sure the cache of child loggers does not prevent the garbage collection of the loggers. If we don't use weak references, there's a small chance of memory leak (we can remove support for weak refs now and add it back when we support only Node 14+). Another option is not to cache them in the first place.

Thoughts?

ullasholla commented:
> Reviewed:
> 
>     * [x]  understandability of the code - nice, clean structure, however a little more complex than I expected

Thanks @tobiasqueck. Some of the complexity was due to caching. I've removed it. The caching covered a niche usecase and I think the tradeoff of having to worry about memory leaks (and having to use weak references is not worth it). Let me know if there are other areas where the code seems more complicated than necessary. We can think of refactoring those areas.

The PR tests now pass (they failed because they run on Node 12 but I developed locally using Node 14. I've removed code that doesn't work with Node 12 and have switched to using Node 12 locally as well).

I have addressed @devinea's comments (I have not marked all the conversations as resolved. I started doing that and then realized that it may be easier for to track the changes if I added a comment but left them as-is. Let me know what works better and I'll keep that in mind for the future).

devinea reviewed (dismissed):
thanks @ullasholla
